File: api/http/java/example/Benchmark.java
/*=============================================================================
# Filename: Benchmark.java
# Author: suxunbin
# Last Modified: 2019-5-17 10:51
# Description: a simple example of multi-thread query
=============================================================================*/

import java.io.*;
import java.util.*;
import jgsc.GstoreConnector;

// before you run this example, make sure that you have started up ghttp service (using bin/ghttp db_name port)
// default db_name: lubm(must be built in advance)
class MyThread extends Thread {
    public static String IP = "127.0.0.1";
    public static int Port = 9000;
    public static String username = "root";
    public static String password = "123456";
    public static boolean correctness = true;
    private int num = 0;
    private int rnum = 0;
    private String sparql;
    private String filename;
    private String RequestType;

    public MyThread(int _num, int _rnum, String _sparql, String _filename, String _RequestType) {
        num = _num;
        rnum = _rnum;
        sparql = _sparql;
        filename = _filename;
        RequestType = _RequestType;
    }
    public void run() {
        // query
        GstoreConnector gc = new GstoreConnector(IP, Port, username, password);
        String res = gc.query("lubm", "json", sparql, RequestType);

        // fquery
        /*GstoreConnector gc = new GstoreConnector(IP, Port, username, password);
        gc.fquery("lubm", "json", sparql, filename, RequestType);
        String res = "";
        FileInputStream in = null;
        try {
            in = new FileInputStream(filename);
            int size = in.available();
            byte[] buffer = new byte[size];
            in.read(buffer);
            in.close();
            res = new String(buffer);
        } catch (IOException e) {
            System.out.println("error occurs: fail to open " + filename);
        }*/

        // count the num
        int m = 0;
        for (int i = 0; i < sparql.length(); ++i) {
            if (sparql.charAt(i) == '?')
                ++m;
            if (sparql.charAt(i) == '{')
                break;
        }
        int n = 0;
        for (int i = 0; i < res.length(); ++i) {
            if (res.charAt(i) == '{')
                ++n;
        }
        int Num = (n - 3) / (m + 1);

        // compare the result
        if (rnum != Num) {
            correctness = false;
            System.out.println("sparql: " + sparql);
            System.out.println("Num: " + Num);
        }
    }
}

public class Benchmark {
    public static void main(String[] args) {
        int tnum = 1000;
        String RequestType = "POST";

        int[] result = {15, 0, 828, 27, 27, 5916};
        String[] sparql = new String[6];

        sparql[0] = "select ?x where"
                    + "{"
                    + "?x <ub:name> <FullProfessor0> ."
                    + "}";
        sparql[1] = "select distinct ?x where"
                    + "{"
                    + "?x <rdf:type>  <ub:GraduateStudent>."
                    + "?y <rdf:type>  <ub:GraduateStudent>."
                    + "?z <rdf:type>  <ub:GraduateStudent>."
                    + "?x <ub:memberOf>  ?z."
                    + "?z <ub:subOrganizationOf> ?y."
                    + "?x <ub:undergaduateDegreeFrom> ?y."
                    + "}";
        sparql[2] = "select distinct ?x where"
                    + "{"
                    + "?x   <rdf:type>  <ub:Course>."
                    + "?x   <ub:name>   ?y."
                    + "}";
        sparql[3] = "select ?x where"
                    + "{"
                    + "?x   <rdf:type>  <ub:UndergraduateStudent>."
                    + "?y   <ub:name>  <Course1>."
                    + "?x   <ub:takesCourse>  ?y."
                    + "?z   <ub:teacherOf>  ?y."
                    + "?z   <ub:name>  <FullProfessor1>."
                    + "?z   <ub:worksFor>   ?w."
                    + "?w   <ub:name>    <Department0>."
                    + "}";
        sparql[4] = "select distinct ?x where"
                    + "{"
                    + "?x   <rdf:type>  <ub:UndergraduateStudent>."
                    + "?y   <ub:name>   <Course1>."
                    + "?x   <ub:takesCourse>  ?y."
                    + "?z   <ub:teacherOf>  ?y."
                    + "?z   <ub:name>  <FullProfessor1>."
                    + "?z   <ub:worksFor>  ?w."
                    + "?w   <ub:name>  <Department0>."
                    + "}";
        sparql[5] = "select distinct ?x where"
                    + "{"
                    + "?x    <rdf:type>    <ub:UndergraduateStudent>."
                    + "}";

        MyThread[] qt = new MyThread[tnum];
        for (int i = 0; i < tnum; ++i) {
            String filename = "result/res" + i + ".txt";
            qt[i] = new MyThread(i, result[i % 6], sparql[i % 6], filename, RequestType);
            qt[i].start();
        }
        for (int i = 0; i < tnum; ++i) {
            try {
                qt[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        if (MyThread.correctness == true)
            System.out.println("The answers are correct!");
        else
            System.out.println("The answers exist errors!");
    }
}

File: api/http/php/example/Benchmark.php
<?php
/*
# Filename: Benchmark.php
# Author: suxunbin
# Last Modified: 2019-5-16 16:05
# Description: a simple example of multi-thread query
 */

// before you run this example, make sure that you have started up ghttp service (using bin/ghttp db_name port)
// default db_name: lubm(must be built in advance)

require '../src/GstoreConnector.php';

$tnum = 1000;
$RequestType = "POST";

$threads = array();
$result = array(15, 0, 828, 27, 27, 5916);
$sparql = array();
$sparql0 = "select ?x where{?x   <ub:name> <FullProfessor0> .}";
$sparql1 = "select distinct ?x where{?x <rdf:type>  <ub:GraduateStudent>.?y <rdf:type>  <ub:GraduateStudent>.?y <rdf:type>  <ub:GraduateStudent>.?x <ub:memberOf>  ?z.?z <ub:subOrganizationOf> ?y.?x <ub:undergaduateDegreeFrom> ?y.}";
$sparql2 = "select distinct ?x where{?x   <rdf:type>  <ub:Course>.?x   <ub:name>   ?y.}";
$sparql3 = "select ?x where{?x   <rdf:type>  <ub:UndergraduateStudent>.?y   <ub:name>  <Course1>.?x   <ub:takesCourse>  ?y.?z   <ub:teacherOf>  ?y.?z   <ub:name>  <FullProfessor1>.?z   <ub:worksFor>   ?w.?w   <ub:name>    <Department0>.}";
$sparql4 = "select distinct ?x where{?x   <rdf:type>  <ub:UndergraduateStudent>.?y   <ub:name>   <Course1>.?x   <ub:takesCourse>  ?y.?z   <ub:teacherOf>  ?y.?z   <ub:name>  <FullProfessor1>.?z   <ub:worksFor>  ?w.?w   <ub:name>  <Department0>.}";
$sparql5 = "select distinct ?x where{?x    <rdf:type>    <ub:UndergraduateStudent>.}";

// thread function
class Mythread extends Thread {
    var $rnum, $sparql, $filename, $RequestType;
    public $correctness;
    public function __construct($rnum, $sparql, $filename, $RequestType) {
        $this->rnum = $rnum;
        $this->sparql = $sparql;
        $this->filename = $filename;
        $this->RequestType = $RequestType;
    }
    public function run () {
        // query
        $gc = new GstoreConnector("127.0.0.1", 9000, "root", "123456");
        $res = $gc->query("lubm", "json", $this->sparql, $this->RequestType);

        // fquery
        //$gc = new GstoreConnector("127.0.0.1", 9000, "root", "123456");
        //$gc->fquery("lubm", "json", $this->sparql, $this->filename, $this->RequestType);
        //$f = fopen($this->filename, "r") or exit("fail to open " . $this->filename);
        //$res = "";
        //while(!feof($f)) {
        //    $res .= (fgets($f). PHP_EOL);
        //}
        //fclose($f);

        // count the num
        $m = 0;
        for ($i = 0; $i < strlen($this->sparql); $i++) {
            if ($this->sparql[$i] == "?")
                $m = $m + 1;
            if ($this->sparql[$i] == "{")
                break;
        }
        $n = 0;
        for ($i = 0; $i < strlen($res); $i++) {
            if ($res[$i] == "{")
                $n = $n + 1;
        }
        $Num = ($n-3)/($m+1);

        // compare the result
        if ($this->rnum != $Num) {
            $this->correctness = false;
            echo "sparql: " . $this->sparql . PHP_EOL;
            echo "Num: " . strval($Num) . PHP_EOL;
        }
        else
            $this->correctness = true;
    }
}

// create sparql
array_push($sparql, $sparql0, $sparql1, $sparql2);
array_push($sparql, $sparql3, $sparql4, $sparql5);

// create the threads
for ($i = 0; $i < $tnum; $i++) {
    $filename = "result/res" . strval($i) . ".txt";
    $threads[] = new Mythread($result[$i%6], $sparql[$i%6], $filename, $RequestType);
}

// start threads
foreach ($threads as $i)
    $i->start();

$correctness = true;
// wait for the threads
foreach ($threads as $i)
{
    $i->join();
    $correctness = $correctness && $i->correctness;
}

if ($correctness == true)
    echo "The answers are correct!" . PHP_EOL;
else
    echo "The answers exist errors!" . PHP_EOL;

?>
